### PR TITLE
Use ProtoEqual matcher in EXPECTS

### PR DESCRIPTION
**Pull request by gdbelvin**
Protobuf does not play well with Go's DeepEqual equality testing that is
used in GoMock.  This PR creates a Protobuf specific equality tester
that uses proto.Equals underneath to improve code stability and correctness.

### CODE REVIEW

File: testonly/matchers/proto.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/mock/gomock"
+)
+
+type protoEqual struct {
+	msg proto.Message
+}
+
+func ProtoEqual(m proto.Message) gomock.Matcher {
```
Comment (RJPercival) [inline testonly/matchers/proto.go]: The linter will insist on documentation for this func.

File: testonly/matchers/atleast.go [hunk 2]
```diff
@@ -21,15 +21,15 @@ import (
 	"github.com/golang/mock/gomock"
 )
 
+type atLeastMatcher struct {
+	num int
+}
```
Comment (RJPercival) [inline testonly/matchers/atleast.go]: I actually prefer the old ordering, where the only exported thing in the file is at the top.

File: testonly/matchers/proto.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	return &protoEqual{msg: m}
+}
+
+func (r *protoEqual) Matches(msg interface{}) bool {
+	m, ok := msg.(proto.Message)
+	if !ok {
+		return false
+	}
+	return proto.Equal(m, r.msg)
```
Comment (codingllama) [inline testonly/matchers/proto.go]: Suggestion: `return ok && proto.Equal(m, r.msg)`

File: testonly/matchers/proto.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+type protoEqual struct {
+	msg proto.Message
+}
+
+func ProtoEqual(m proto.Message) gomock.Matcher {
+	return &protoEqual{msg: m}
+}
+
+func (r *protoEqual) Matches(msg interface{}) bool {
```
Comment (codingllama) [inline testonly/matchers/proto.go]: nit: Why `r`? Maybe `m` (for matcher) or `pe` (for ProtoEqual)?

File: testonly/matchers/proto.go (hunk 1 above)
Comment (codingllama) [inline testonly/matchers/proto.go]: Please document.

File: testonly/matchers/proto.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	m, ok := msg.(proto.Message)
+	if !ok {
+		return false
+	}
+	return proto.Equal(m, r.msg)
+}
+
+func (r *protoEqual) String() string {
+	return fmt.Sprintf("is %s", r.msg)
```
Comment (codingllama) [inline testonly/matchers/proto.go]: Suggestion: "is equal to %v", so it looks like the regular gomock.Eq message?

https://github.com/golang/mock/blob/master/gomock/matchers.go#L51

File: testonly/matchers/atleast.go [hunk 6]
```diff
@@ -21,15 +21,15 @@ import (
 	"github.com/golang/mock/gomock"
 )
 
+type atLeastMatcher struct {
```
Comment (codingllama) [inline testonly/matchers/atleast.go]: Please revert the changes to this file. There's no need to touch it in this PR.

File: testonly/matchers/proto.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+func (r *protoEqual) Matches(msg interface{}) bool {
+	m, ok := msg.(proto.Message)
+	if !ok {
+		return false
+	}
+	return proto.Equal(m, r.msg)
+}
+
+func (r *protoEqual) String() string {
```
Comment (RJPercival) [inline testonly/matchers/proto.go]: Doesn't need a pointer receiver, since the only thing in the struct is a single pointer so it's very cheap to copy.

File: testonly/matchers/proto.go (hunk 5 above)
Comment (gdbelvin) [inline testonly/matchers/proto.go]: done

File: testonly/matchers/proto.go (hunk 3 above)
Comment (gdbelvin) [inline testonly/matchers/proto.go]: I think the expanded form is slightly more canonical. I borrowed it from the [gRPC documenation](https://github.com/grpc/grpc-go/blob/master/Documentation/gomock-example.md)

File: testonly/matchers/proto.go (hunk 4 above)
Comment (gdbelvin) [inline testonly/matchers/proto.go]: done

File: testonly/matchers/proto.go (hunk 1 above)
Comment (gdbelvin) [inline testonly/matchers/proto.go]: done

Comment (gdbelvin) [inline testonly/matchers/proto.go]: done

File: testonly/matchers/atleast.go (hunk 2 above)
Comment (gdbelvin) [inline testonly/matchers/atleast.go]: I'm used to seeing types, followed by their methods. But happy to revert.

File: testonly/matchers/atleast.go (hunk 6 above)
Comment (gdbelvin) [inline testonly/matchers/atleast.go]: ack

**Comment by RJPercival** [comment]
As simple as this matcher is, it arguably should have a small test since the AtLeast matcher does.

File: testonly/matchers/atleast.go (hunk 2 above)
Comment (RJPercival) [inline testonly/matchers/atleast.go]: The `ProtoEqual()` and `AtLeast()` funcs aren't methods though.

**Comment by gdbelvin** [comment]
Added test

File: testonly/matchers/proto_test.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package matchers
+
+import (
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
```
Comment (RJPercival) [inline testonly/matchers/proto_test.go]: s/gogo/golang/

File: testonly/matchers/proto_test.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	for _, tc := range []struct {
+		a, b  proto.Message
+		equal bool
+	}{
+		{a: &trillian.SignedLogRoot{}, b: &trillian.SignedLogRoot{}, equal: true},
+		{a: &trillian.SignedLogRoot{}, b: &trillian.SignedLogRoot{RootHash: []byte{}}, equal: true},
+		{a: &trillian.SignedLogRoot{}, b: &trillian.SignedLogRoot{RootHash: []byte{0x01}}, equal: false},
+	} {
+		if got, want := ProtoEqual(tc.a).Matches(tc.b), tc.equal; got != want {
+			t.Errorf("ProtoEqual(%s).Matches(%s): %v, want %v", tc.a, tc.b, got, want)
```
Comment (RJPercival) [inline testonly/matchers/proto_test.go]: s/:/ =/

File: testonly/matchers/proto.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package matchers
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
```
Comment (RJPercival) [inline testonly/matchers/proto.go]: s/gogo/golang/

Comment (gdbelvin) [inline testonly/matchers/proto.go]: done

File: testonly/matchers/proto_test.go (hunk 8 above)
Comment (gdbelvin) [inline testonly/matchers/proto_test.go]: done

File: testonly/matchers/proto_test.go (hunk 9 above)
Comment (gdbelvin) [inline testonly/matchers/proto_test.go]: done
